File: java-patterns/src/main/java/behavioral/mediator/Mediator.java
package behavioral.mediator;

/**
 * Usage: 中介接口 <br/>
 * Date: 2023/5/13 16:34 <br/>
 *
 * @author <a href="mailto:hanzhang2566@foxmail.com">hanzhang</a>
 */
public interface Mediator {
    void register(Colleague colleague);

    void relay(Colleague colleague);

}

File: go-patterns/behavioral/mediator/mediator.go
package mediator

type Mediator interface {
	register(c Colleague)
	relay(c Colleague)
}

type FooMediator struct {
	colleagues []Colleague
}

func (fm *FooMediator) register(c Colleague) {
	if contains := fm.contains(c); contains {
		return
	}
	fm.colleagues = append(fm.colleagues, c)
	c.setMedium(fm)
}

func (fm *FooMediator) relay(c Colleague) {
	for _, colleague := range fm.colleagues {
		if colleague == c {
			continue
		}
		colleague.receive()
	}
}

func (fm *FooMediator) contains(c Colleague) bool {
	for _, colleague := range fm.colleagues {
		if colleague == c {
			return true
		}
	}
	return false
}
